### PR TITLE
raftstore: fix possiblely calling the wrong callback

### PR DESCRIPTION
**Pull request by BusyJay**
When a follower with uncommitted log becomes leader and then propose a message immediately, the callback of the message will be considered as stale callback, which apparently is not.

This pr will take term into consideration, and hold the callback queue when applied entry's term is less than the callback's term.

@ngaut @siddontang @disksing PTAL

### CODE REVIEW

File: tests/raftstore/transport_simulate.rs [hunk 1]
```diff
@@ -215,11 +220,19 @@ impl FilterFactory for Isolate {
...
+///
+/// If `msg_type` is None, all message will be filtered.
+/// If direction is 1, only output messages will be fitered;
+/// if direction is 2, only input messages will be filtered;
+/// if direction is 3, both input and output messages will be filtered.
 pub struct FilterRegionPacket {
     region_id: u64,
     store_id: u64,
     drop: AtomicBool,
+    direction: u8,
```
Comment (ngaut) [inline tests/raftstore/transport_simulate.rs]: Use enum instead of u8.

File: tests/raftstore/transport_simulate.rs [hunk 2]
```diff
@@ -215,11 +220,39 @@ impl FilterFactory for Isolate {
...
+            Direction::Send | Direction::Both => true,
+            Direction::Recv => false,
+        }
+    }
+}
+
+/// Drop specified messages for the store with special region.
+///
+/// If `msg_type` is None, all message will be filtered.
 pub struct FilterRegionPacket {
```
Comment (siddontang) [inline tests/raftstore/transport_simulate.rs]: I see that for all Filter and its corresponding FilterFactory, they have nearly same structure fields and same initialization function, too redundant. Should we think a better way to simplify these? 

/cc @ngaut @BusyJay

File: tests/raftstore/server.rs [hunk 3]
```diff
@@ -250,9 +250,14 @@ impl Simulator for ServerCluster {
         Ok(())
     }
 
-    fn hook_transport(&self, node_id: u64, filters: Vec<Box<Filter>>) {
+    fn add_filter(&self, node_id: u64, filters: Box<Filter>) {
```
Comment (siddontang) [inline tests/raftstore/server.rs]: s/filters/filter/

File: tests/raftstore/cluster.rs [hunk 4]
```diff
@@ -51,7 +51,8 @@ pub trait Simulator {
     fn call_command(&self, request: RaftCmdRequest, timeout: Duration) -> Result<RaftCmdResponse>;
     fn send_raft_msg(&self, msg: RaftMessage) -> Result<()>;
     fn get_store_sendch(&self, node_id: u64) -> Option<SendCh>;
-    fn hook_transport(&self, node_id: u64, filters: Vec<Box<Filter>>);
+    fn add_filter(&self, node_id: u64, filters: Box<Filter>);
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: s/filters/filter/

File: tests/raftstore/transport_simulate.rs [hunk 5]
```diff
@@ -78,8 +79,12 @@ impl<T: Transport> SimulateTransport<T> {
         }
     }
 
-    pub fn set_filters(&mut self, filters: Vec<Box<Filter>>) {
-        self.filters = filters;
+    pub fn clear_filter(&mut self) {
+        self.filters.clear();
+    }
+
+    pub fn add_filters(&mut self, filter: Box<Filter>) {
```
Comment (siddontang) [inline tests/raftstore/transport_simulate.rs]: add_filter

**Comment by siddontang** [comment]
Rest LGTM

File: tests/raftstore/test_multi.rs [hunk 6]
```diff
@@ -267,3 +268,85 @@ fn test_multi_server_random_restart() {
...
+    cluster.add_filter(IsolateRegionStore::new(1, 1)
+        .msg_type(MessageType::MsgAppend)
+        .direction(Direction::Recv));
+    // Make 2 have no way to know the uncommitted entries can be applied when it's still follower.
+    cluster.add_filter(IsolateRegionStore::new(1, 2)
+        .msg_type(MessageType::MsgHeartbeat)
+        .direction(Direction::Recv));
+    cluster.must_put(b"k2", b"v2");
+
+    // 1 must commit, but 2 is not.
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: k1 ?

File: tests/raftstore/test_multi.rs [hunk 7]
```diff
@@ -267,3 +268,85 @@ fn test_multi_server_random_restart() {
...
+        let engine = cluster.get_engine(i);
+        must_get_equal(&engine, b"k1", b"v1");
+    }
+
+    let engine3 = cluster.get_engine(3);
+    must_get_none(&engine3, b"k1");
+
+    // now only 1 and 2 can step to leader.
+
+    // hack: first append will append log, second append will set commit,
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: first message will be append to log, second message will be set to committed

File: tests/raftstore/test_multi.rs [hunk 8]
```diff
@@ -267,3 +268,85 @@ fn test_multi_server_random_restart() {
...
+    // So only allowing first append to make 2 have uncommitted entries.
+    cluster.add_filter(IsolateRegionStore::new(1, 2)
+        .msg_type(MessageType::MsgAppend)
+        .direction(Direction::Recv)
+        .allow(1));
+    // Make 2 have no way to know the uncommitted entries can be applied when it becomes leader.
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: What does 2 mean ?

File: tests/raftstore/test_multi.rs [hunk 9]
```diff
@@ -267,3 +268,85 @@ fn test_multi_server_random_restart() {
...
+    // node 1 and node 2 must have k2, but node 3 must not.
+    for i in 1..3 {
+        let engine = cluster.get_engine(i);
+        must_get_equal(&engine, b"k1", b"v1");
+    }
+
+    let engine3 = cluster.get_engine(3);
+    must_get_none(&engine3, b"k1");
+
+    // now only 1 and 2 can step to leader.
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: What does 1 or 2 mean ?

File: tests/raftstore/test_multi.rs [hunk 10]
```diff
@@ -267,3 +268,88 @@ fn test_multi_server_random_restart() {
...
+fn test_leader_change_with_uncommitted_log<T: Simulator>(cluster: &mut Cluster<T>) {
+    cluster.cfg.store_cfg.raft_election_timeout_ticks = 50;
+    // We use three peers([1, 2, 3]) for this test.
+    cluster.run();
+
+    sleep_ms(500);
+
+    // guarantee peer 1 is leader
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+    cluster.must_put(b"k0", b"v0");
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: I think you can remove the two lines 281-282.

File: tests/raftstore/test_multi.rs [hunk 11]
```diff
@@ -267,3 +268,88 @@ fn test_multi_server_random_restart() {
...
+        let engine = cluster.get_engine(i);
+        must_get_equal(&engine, b"k1", b"v1");
+    }
+
+    let engine3 = cluster.get_engine(3);
+    must_get_none(&engine3, b"k1");
+
+    // now only peer 1 and peer 2 can step to leader.
+
+    // hack: first MsgAppend will append log, second MsgAppend will set commit index,
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: s/will/will be/

File: tests/raftstore/test_multi.rs [hunk 12]
```diff
@@ -267,3 +268,88 @@ fn test_multi_server_random_restart() {
...
+    // So only allowing first MsgAppend to make 2 have uncommitted entries.
```
Comment (ngaut) [inline tests/raftstore/test_multi.rs]: s/2/peer 2/

File: tests/raftstore/cluster.rs [hunk 13]
```diff
@@ -513,10 +519,10 @@ impl<T: Simulator> Cluster<T> {
         }
     }
 
-    pub fn reset_transport_hooks(&mut self) {
+    pub fn clear_filter(&mut self) {
```
Comment (ngaut) [inline tests/raftstore/cluster.rs]: s/clear_filter/clear_filters/

**Comment by ngaut** [comment]
LGTM

File: tests/raftstore/test_multi.rs [hunk 14]
```diff
@@ -267,3 +268,88 @@ fn test_multi_server_random_restart() {
...
 }
+
+fn test_leader_change_with_uncommitted_log<T: Simulator>(cluster: &mut Cluster<T>) {
+    cluster.cfg.store_cfg.raft_election_timeout_ticks = 50;
+    // disable compact log to make test more stable.
+    cluster.cfg.store_cfg.raft_log_gc_threshold = 1000;
+    // We use three peers([1, 2, 3]) for this test.
+    cluster.run();
+
+    sleep_ms(500);
```
Comment (tiancaiamao) [inline tests/raftstore/test_multi.rs]: what's the purpose of this line?

Comment (BusyJay) [inline tests/raftstore/test_multi.rs]: Because election timeout has been enlarged, so here wait a little time in case CI is very slow.

**Comment by tiancaiamao** [comment]
LGTM
/ref https://github.com/pingcap/tikv/issues/489

**Pull request merged**